shell/systemui-extra/liblingmosdk-base/man/timer/libkytimer.3
.TH "LIBKYTIMER" 3 "Fri Sep 22 2023" "My Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
libkytimer \- LINGMOSDK C语言定时器模块  

.SH SYNOPSIS

.SS "Classes"

.SS "Typedefs"

.in +1c
.ti -1c
.RI "typedef void *(* \fBtime_handler\fP) (void *user_data)"
.br
.ti -1c
.RI "typedef struct \fB_KTimerNode\fP \fBKTimerNode\fP"
.br
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBKTimerAttribute\fP { \fBKTIMER_SINGLESHOT\fP = 0, \fBKTIMER_PERIODIC\fP = 1, \fBKTIMER_NEVER\fP = 2 }"
.br
.ti -1c
.RI "enum \fBKTimerType\fP { \fBKTIMER_ABSOLUTE\fP = 0, \fBKTIMER_RELATIVE\fP }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "int \fBkdk_timer_init\fP ()"
.br
.RI "初始化定时器核心组件 "
.ti -1c
.RI "size_t \fBkdk_timer_start\fP (unsigned int intervalms, time_handler callback, KTimerAttribute attr, KTimerType type, void *userdata, int freeOnDelete)"
.br
.RI "启动一个定时器 "
.ti -1c
.RI "void \fBkdk_timer_stop\fP (size_t timerfd)"
.br
.RI "停止给定的定时器 "
.ti -1c
.RI "void \fBkdk_timer_reset\fP (size_t timerfd, unsigned int intervalms)"
.br
.RI "重置定时器时间至intervalms，以毫秒为单位 "
.ti -1c
.RI "void \fBkdk_timer_destroy\fP ()"
.br
.RI "销毁定时器核心 "
.in -1c
.SH DATA STRUCTURES

\fBtypedef struct _KTimerNode {
.br
    pthread_mutex_t	lock;
	size_t			fd;			//序列号
	time_handler	callback;	//到期后执行的函数
	int				freeOnDelete;	//删除定时器时自动释放userdata
	void*			userdata;	//callback使用的参数，外部使用alloc分配内存，在delete定时器的时候会根据配置自动释放
	unsigned int	intervalms;	//定时器间隔时间，单位毫秒
	KTimerAttribute	attr;		//触发类型，单次触发或多次触发
	KTimerType		type;		//时钟类型，绝对时间还是相对时间
	struct _KTimerNode*	next;
.br
} KTimerNode;\fP

There are other members not documented here, which are only meant for
libkytimer internal use.
.SH "Detailed Description"
.PP 
LINGMOSDK C语言定时器模块 
.SH "CONFORMING TO"
These functions are as per the withdrawn POSIX.1e draft specification.
The following functions are Linux extensions:
.BR kdk_timer_init (),
.BR kdk_timer_start (),
.BR kdk_timer_stop (),
.BR kdk_timer_reset ()
and
.BR kdk_timer_destroy ().
.SH "SEE ALSO"
.BR kdk_timer_init (3),
.BR kdk_timer_start (3),
.BR kdk_timer_stop (3),
.BR kdk_timer_reset (3)
and
.BR kdk_timer_destroy (3).
.SH "Author"
.PP 
Generated automatically by Doxygen for libkytimer.h from the source code\&.


CHANGES
2024.02.4, released July 13th, 2024

	Important / security related fixes.

	Infrastructure: Ensure a custom tar program specified with
	make TAR=/path/to/custom/tar is used everwhere.

	Defconfig: Pine64 rockpro64: Build ethernet driver in kernel
	rather than as module.

	Updated/fixed packages: apache, composer, daq, fluent-bit,
	ghostscript, glibc, liblockfile, libmad, libopenssl, mesa3d,
	network-manager, ntp, octave, openssh, openvpn, php, python3,
	qt6base, raspberrypi-usbboot, redis, sentry-cli, systemd, tor,
	vlc, xmrig, znc

2024.02.3, released June 10th, 2024

	Important / security related fixes.

	Misc:
	- Fixed a number of compilation issues for hosts using GCC
          14.x, including for kconfig.

	Download:
	- Drop redundant --passive-ftp from default wget options and
	  migrate a number of packages using FTP upstreams to HTTP(S)
	  for compatibility with wget2 which does not have FTP support

	- Ensure downloads are correctly shared for host packages that
	  try to do so using <pkg>_DL_SUBDIR and ensure the correct
	  toplevel directory name is used in generated tarballs when
	  downloads are shared.

	Arch: Enable SSE2 support for x86 Pentium M variant.

	Defconfigs: Orangepi PC/R1, Raspberrypi zero2w, Sipeed
	licheepi nano, Wandboard: Fix compilation issues. PC x86-64
	bios/efi: add wireless regdb to fix kernel warning during
	boot, iMX.8 boards: Adjust FIT external data position to match
	imx-mkimage changes, fixing boot issues.

	Updated/fixed packages: am33x-cm, apcupsd,
	arm-trusted-firmware, binutils-bare-metal, bpftool, catch2,
	ccache, cjson, clamav, cog, coreutils, cryptodev-linux,
	dhcpcd, dmraid, dnsmasq, docker-cli, docker-compose,
	docker-engine, doom-wad, duma, elf2flt, ebtables, elfutils,
	evemu, file, fluidsynth, freeradius-client, freeradius-server,
	freerdp, fwts, gcc, gcc-bare-metal, git, gnuplot, go,
	gr-osmosdr, gst1-plugins-good, hidapi, imagemagick,
	intel-gmmlib, ivi-homescreen, janus-gateway, kexec, kmod,
	kodi, libarchive, libdrm, libedit, libffi, libglib2, libhtp,
	libjwt, libkrb5, libmanette, libmdbx, libmnl,
	libnetfilter_acct, libnetfilter_log, libnpth, libopenssl,
	libpng, libuhttpd, liburiparser, libutempter, libvncserver,
	libwpe, lightning, linux-firmware, linux-pam, lua-cffi,
	luajit, lvm2, lynx, mariadb, mesa3d, micropython, minisign,
	musl, mutt, ncftp, network-manager, newlib-bare-metal, nfacct,
	nfs-utils, octave, openjpeg, openocd, oprofile,
	perl-dbd-mysql, php, postgresql, pppd, privoxy, procps-ng,
	proftpd, proj, putty, python-ml-dtypes, python-pillow,
	python-rds-py, qemu, qt5base, redis-plus-plus, rtl8189fs,
	ruby, samba4, shadow, smcroute, speech, squid, sscep,
	suricata, swupdate, systemd, ti-sgx-km, tinycbor, tinyproxy,
	tor, tpm2-tss, ttyd, uboot-tools, ulogd, vala, weston,
	wpewebkit, x11vnc, zbar, zeromq, zfs, zip, zxing-cpp

	Issues resolved (http://bugs.uclibc.org):

        #15766: LINGMO_PACKAGE_WESTON_XWAYLAND -- 12.0.1 -- fails to build..
	#15973: imx8m bootloader image generation broken since update..
        #16075: Cannot use linux kernel extensions from external tree..
	#16093: kmod and gcc-14 build error

2024.02.2, released May 7th, 2024

	Important / security related fixes.

        Security: The /dev/shm entry in the default /etc/fstab used
        for busybox and sysv init systems was missing the sticky bit,
        allowing users to delete and replace other users files.

        Defconfigs: Mangopi mq1rdw1: Lock kernel headers to 6.6.x.

	Updated/fixed packages: apache, botan,
	brcmfmac_sdio-firmware-rpi, clamav, cmake, containerd,
	crucible, dmenu-wayland, dmraid, docker-cli, docker-engine,
	domoticz, fmc, freeipmi, gdal, gdb, giflib, gnupg2, go,
	google-breakpad, googlefontdirectory, imx-gpu-viv, inih,
	libassuan, libblockdev, libcurl, libdrm, libgpg-error,
	libksba, libnpth, libressl, libtracefs, lxc, mbedtls, mesa3d,
	oatpp, opensc, oprofile, perl-net-ssleay, petitboot,
	php-xdebug, psplash, python-pypng, python-segno,
	python-service-identity, python3, skeleton-init-sysv, systemd,
	tcf-agent, tiff, uuu, wolfssl, xmrig, zic

	Issues resolved (http://bugs.uclibc.org):

        #14416: Mesa3D missing DRI3 support for various Gallium drivers
        #16009: 'select' does not work with 'choice'
        #16036: android-tools version is too old

2024.02.1, released March 26th, 2024

	Important / security related fixes.

	Toolchain: Dropped old Codesourcery ARM / AARCH64 and
	Codescape MIPS external toolchains as they use old Glibc
	versions not compatible with 64-bit time_t

	Defconfigs: Starfive visionfive2: Use correct genimage
	configuration. TI am62x_sk / am64_sk: Use correct optee-os
	platform setting and use a fixed ti-k3-r5-loader version.

	Infrastructure: Use git to fetch CVE data from the FKIE
	repository to work around an issue with missing meta files in
	the releases.

	Updated/fixed packages: axel, bash, bitcoin, botan, busybox,
	containerd, cryptsetup, cvs, docker-compose, docker-engine,
	ell, expat, giflib, gmp, go, harfbuzz, intel-gmmlib,
	intel-mediadriver, iwd, kodi, kodi-pvr-hts, libcgroup, libdrm,
	libgit2, libkrb5, libnfs, libunwind, libvpl,
	libtorrent-rasterbar, libxml2, linux, mariadb, mesa3d,
	mmc-utils, msmtp, multipath-tools, neon, onevpl-intel-gpu,
	openssh, openvpn, osm2pgsql, owfs, php, poco, policycoreutils,
	postgis, privoxy, pulseaudio, pure-ftpd, putty,
	python-configshell-fb, python-rtslib-fb, rauc, restorecond,
	spice, squid, targetcli-fb, tzdata, util-linux, vdr, xmrig,
	xvisor, zic

	Issues resolved (http://bugs.uclibc.org):

	#15997: Build Squid with GCC12.3 in Buildroot 2024.02 will fail

2024.02, released March 5th, 2024

	Various fixes.

	Updated/fixed packages: botan, conmon, dash, efivar, expat,
	fluent-bit, gcc, gst1-vaapi, libcoap, libselinux, liburing,
	libva, libxml2, ltp-testsuite, luvi, newlib-bare-metal,
	nginx-naxsi, octave, openblas, openvmtools, parted, poco,
	powertop, python-aiohttp, python-cheroot, python-django,

shell/systemui-extra/liblingmosdk-base/src/timer/libkytimer.c
/*
 * Copyright (C) 2023, KylinSoft Co., Ltd.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Authors: Yunhe Liu <liuyunhe@kylinos.cn>
 *
 */


#include <stdio.h>
#include <string.h>
#include <sys/timerfd.h>
#include <time.h>
#include <pthread.h>
#include <stdint.h>
#include <poll.h>
#include <errno.h>
#include <unistd.h>
#include <sys/epoll.h>
#include "libkytimer.h"
#include "sdkmarcos.h"

static void* timerCoreThread(void* data);		//定时器循环核心
static KTimerNode* findNodeByFD(unsigned int fd);	//根据fd查找节点
static KTimerNode* newTimerNode();
static void freeTimerNode(KTimerNode* node);
static int	testNodeInList(KTimerNode* node);
static void	insertIntoList(KTimerNode* node);	//插入节点，按照间隔时间的顺序从小到大排序
static void	deleteFromList(KTimerNode* node , int locked);		//从链表中删除节点
static pthread_t	g_coreThreadID;		//核心循环线程
static KTimerNode*	g_list	= NULL;
static int	epollfd;	//事件池句柄
static int	curTimerCounts;	//当前已注册定时器数量

int	kdk_timer_init();	//初始化定时器核心
size_t 	kdk_timer_start(unsigned int intervalms, time_handler callback, KTimerAttribute attr, KTimerType type, void* userdata, int freeOnDelete);	//注册并开始一个定时器，返回定时器句柄号
void	kdk_timer_stop(size_t timerfd);	//停止一个定时器
void	kdk_timer_reset(size_t timerfd , unsigned int intervalms);
void	kdk_timer_destroy() DESTRUCTOR;		//销毁定时器核心

#ifndef	TFD_TIMER_CANCEL_ON_SET
#define	TFD_TIMER_CANCEL_ON_SET	(1 << 1)
#endif

int kdk_timer_init()
{
	if (g_coreThreadID > 0)
		return 0;

	epollfd	= epoll_create1(EPOLL_CLOEXEC);
	if (epollfd <= 0)
	{
		printf("Epoll事件池创建失败！%s\n" , strerror(errno));
		return errno;
	}

	if (pthread_create(&g_coreThreadID , NULL , timerCoreThread , NULL))
	{
		printf("kyTimer 定时器核心初始化失败：%s\n" , strerror(errno));
		return errno;
	}
	return 0;
}

size_t kdk_timer_start(unsigned int intervalms, time_handler callback, KTimerAttribute attr, KTimerType type, void* userdata, int freeOnDelete)
{
	if (g_coreThreadID <= 0 && kdk_timer_init())
	{
		printf("定时器注册失败：定时器全局句柄初始化失败！\n");
		return 0;
	}
	if (curTimerCounts >= KTIMER_MAXTIMERFD)
	{
		printf("定时器注册失败：定时器注册数量已超限！\n");
		return 0;
	}

	KTimerNode* node	= newTimerNode();
	if (!node)
	{
		printf("定时器注册失败：无法分配内存，%s\n" , strerror(errno));
		return 0;
	}

	node->callback	= callback;
	node->userdata	= userdata;
	node->intervalms	= intervalms;
	node->attr		= attr;
	node->type		= type;
	node->freeOnDelete	= freeOnDelete;

	if (type == KTIMER_ABSOLUTE)
		node->fd		= timerfd_create(CLOCK_MONOTONIC , TFD_CLOEXEC|TFD_NONBLOCK);
	else
		node->fd		= timerfd_create(CLOCK_REALTIME , TFD_CLOEXEC|TFD_NONBLOCK);

	if ((int)node->fd <= 0)
	{
		printf("定时器注册失败：timerfd_create失败，%s\n" , strerror(errno));
		freeTimerNode(node);
		return 0;
	}
	struct itimerspec tval;
	tval.it_value.tv_sec	= intervalms / 1000;
	tval.it_value.tv_nsec	= (intervalms % 1000) * 1000000;

	if (attr == KTIMER_PERIODIC)	//多次触发时设置interval项
	{
		tval.it_interval.tv_sec	= tval.it_value.tv_sec;
		tval.it_interval.tv_nsec	= tval.it_value.tv_nsec;
	}
	else
	{
		tval.it_interval.tv_sec	= 0;
		tval.it_interval.tv_nsec	= 0;
	}

	timerfd_settime(node->fd, 0, &tval, NULL);

	//注册到Epoll池中
	struct epoll_event ev;
	memset(&ev , 0 , sizeof(struct epoll_event));
	if (attr == KTIMER_SINGLESHOT)
		ev.events	= EPOLLIN | EPOLLONESHOT;
	else
		ev.events	= EPOLLIN;
	ev.data.ptr	= node;
	if (epoll_ctl(epollfd , EPOLL_CTL_ADD , node->fd , &ev))
	{
		printf("定时器注册失败：epoll_ctl错误，%s\n" , strerror(errno));
		freeTimerNode(node);
		return 0;
	}

	insertIntoList(node);

	return node->fd;
}

//停止一个定时器，并从列表中移除
void kdk_timer_stop(size_t timerfd)
{
	if (timerfd <= 0)
		return;
	KTimerNode* node	= findNodeByFD(timerfd);
	if (node)
	{
		if (epoll_ctl(epollfd , EPOLL_CTL_DEL , timerfd , NULL))
		{
			printf("无法注销文件句柄%zd : %s\n" , timerfd , strerror(errno));
			return;
		}
		deleteFromList(node , 0);
	}
	return;
}

//定时器核心线程，负责循环处理已到期的定时器
static void* timerCoreThread(void* data)
{
	pthread_detach(pthread_self());

	KTimerNode* knode	= NULL;
	int read_fds = 0;

	while (1)
	{
		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE , NULL);
		pthread_testcancel();
		pthread_setcancelstate(PTHREAD_CANCEL_DISABLE , NULL);

		struct epoll_event activeEvs[KTIMER_MAXTIMERFD];
		read_fds	= epoll_wait(epollfd , activeEvs , KTIMER_MAXTIMERFD , -1);
		if (read_fds < 0)
		{
			//printf("epoll wait error , %s\n" , strerror(errno));
			continue;
		}
		for (int i = 0 ; i < read_fds ; i ++)
		{
			knode	= activeEvs[i].data.ptr;
			pthread_mutex_lock(&knode->lock);
			if (!testNodeInList(knode))
				continue;
			uint64_t dep;
			size_t rd = read(knode->fd , &dep , sizeof(uint64_t));
			if (rd <= 0)
				continue;

			if (knode && knode->attr != KTIMER_NEVER && knode->callback)
			{
#if 1
				pthread_attr_t attr;
				pthread_t tid;
				pthread_attr_init(&attr);
				pthread_attr_setdetachstate(&attr , PTHREAD_CREATE_DETACHED);
				pthread_create(&tid , &attr , knode->callback , knode->userdata);

				knode->freeOnDelete	= 0;
#else
				knode->callback(knode->userdata);
#endif
			}

			if (knode->attr == KTIMER_SINGLESHOT)	//单次触发的在触发后就从链表中移除
			{
				// FixMe: 如果设置了freeOnDelete，并且采用上面的线程执行模式，
				// 		  则会发生段错误。因此只能在启用线程执行模式的时候强制禁止
				// 		  freeOnDelete。
				deleteFromList(knode , 1);
			}
			if (testNodeInList(knode))
				pthread_mutex_unlock(&knode->lock);
		}
	}

	return NULL;
}

static KTimerNode* findNodeByFD(unsigned int fd)
{
	KTimerNode* tmp	= g_list;
	while (tmp)
	{
		if (tmp->fd == fd)
			return tmp;
		tmp	= tmp->next;
	}
	return NULL;
}

static int testNodeInList(KTimerNode* node)
{
	KTimerNode* pos	= g_list;
	while (pos)
	{
		if (pos == node)
			return 1;
		pos	= pos->next;
	}
	return 0;
}

static void	insertIntoList(KTimerNode* node)
{
	curTimerCounts ++;
	if (!g_list)
	{
		g_list	= node;
	}
	else
	{
		if (g_list->intervalms > node->intervalms)
		{
			node->next	= g_list;
			g_list	= node;
			return;
		}

		KTimerNode* cur		= g_list->next;
		KTimerNode* prev	= g_list;

		while (cur)
		{
			if (cur->intervalms > node->intervalms)
			{
				node->next	= cur;
				prev->next	= node;
				return;
			}
			prev	= cur;
			cur	= cur->next;
		}

		prev->next	= node;
		node->next	= NULL;
	}
	return;
}

static void deleteFromList(KTimerNode* node , int locked)
{
	if (!g_list)
		return;
	curTimerCounts --;
	if (!locked)
		pthread_mutex_lock(&node->lock);
	if (node == g_list)
	{
		g_list	= g_list->next;
		freeTimerNode(node);
		return;
	}

	KTimerNode* pos	= g_list;
	while (pos)
	{
		if (pos->next && pos->next == node)
		{
			pos->next	= pos->next->next;
			freeTimerNode(node);
			return;
		}
		pos	= pos->next;
	}
	return;
}

static KTimerNode* newTimerNode()
{
	KTimerNode *node = (KTimerNode*)calloc(1 , sizeof(KTimerNode));
	if (!node)
		return NULL;
	pthread_mutex_init(&node->lock , NULL);
	return node;
}

static void freeTimerNode(KTimerNode *node)
{
	if (! node)
		return;
	if (node->fd > 0)
		close(node->fd);
	if (node->userdata && node->freeOnDelete)
		free(node->userdata);
	free(node);
}

void kdk_timer_reset(size_t timerfd , unsigned int intervalms)
{
	KTimerNode* node	= findNodeByFD(timerfd);
	if (!node)
		return;
	node->intervalms	= intervalms;
	struct itimerspec tval;
	tval.it_value.tv_sec	= intervalms / 1000;
	tval.it_value.tv_nsec	= (intervalms % 1000) * 1000000;

	if (node->attr == KTIMER_PERIODIC)	//多次触发时设置interval项
	{
		tval.it_interval.tv_sec	= tval.it_value.tv_sec;
		tval.it_interval.tv_nsec	= tval.it_value.tv_nsec;
	}
	else
	{
		tval.it_interval.tv_sec	= 0;
		tval.it_interval.tv_nsec	= 0;
	}
	timerfd_settime(timerfd , 0 , &tval , NULL);
	return;
}

void kdk_timer_destroy()	//停止所有定时器，并销毁定时器核心
{
	//终止线程运行
	if (g_coreThreadID > 0)
		pthread_cancel(g_coreThreadID);

	//销毁链表
	KTimerNode* node	= g_list;
	while (node)
	{
		deleteFromList(node , 0);
		node	= g_list;
	}

	//关闭epoll池
	if (epollfd > 0)
		close(epollfd);
	return;
}
